### ISSUE TITLE
ETag middleware wastefully overwrites ETags

### ISSUE THREAD

**Issue by jamesarosen** [honojs/hono#1195]
The `etag` middleware currently has this logic:

```ts
const res = c.res as Response
const clone = res.clone()
const hash = await sha1(res.body || '')
```

If `c.res` already has an `ETag` header, this wastefully clones the response and computes the SHA1 hash of the body.

**Comment by jamesarosen** [comment]
Proposed change:

```ts
export const etag = (options: ETagOptions = { weak: false }): MiddlewareHandler => {
  return async (c, next) => {
    const ifNoneMatch = c.req.header('If-None-Match') || c.req.header('if-none-match')

    await next()

    const res = c.res as Response
    let etag = c.res.headers.get('etag')
    let unusedRes = res

    if (!etag) {
      unusedRes = res.clone()
      const hash = await sha1(res.body || '')
      etag = options.weak ? `W/"${hash}"` : `"${hash}"`
    }

    if (ifNoneMatch && ifNoneMatch === etag) {
      await unusedRes.blob() // Force using body
      c.res = new Response(null, {
        status: 304,
        statusText: 'Not Modified',
        headers: {
          ETag: etag,
        },
      })
      c.res.headers.delete('Content-Length')
    } else {
      c.res = new Response(unusedRes.body, unusedRes)
      c.res.headers.set('ETag', etag)
    }
  }
}
```

Alternatively, the single `etag` middleware could be split into two:

```ts
/**
 * Ensure the response has an ETag.
 */
export const etag = (options: ETagOptions = { weak: false }): MiddlewareHandler => {
  return async (c, next) => {
    await next()
    const res = c.res as Response
    if (res.headers.has('etag')) return

    const clone = res.clone()
    const hash = await sha1(res.body || '')
    const etag = options.weak ? `W/"${hash}"` : `"${hash}"`
    c.res = new Response(clone.body, clone)
    c.res.headers.set('ETag', etag)
  }
}

export async function ifNoneMatch(c: Context, next: Next) {
  const ifNoneMatchTag = c.req.header('If-None-Match')
  await next()
  const res = c.res as Response
  const etag = res.headers.get('etag')

  if (ifNoneMatchTag && etag && ifNoneMatchTag === etag) {
    c.res = new Response(null, {
      status: 304,
      statusText: 'Not Modified',
      headers: {
        ETag: etag,
      },
    })
    c.res.headers.delete('Content-Length')
  }
}
```

**Comment by yusukebe** [comment]
Hi @jamesarosen 

This will be a good fix! I'd like to go with the above pattern. Could you create the PR? Thanks!

**Comment by jamesarosen** [comment]
Will do 👍

**Comment by jamesarosen** [comment]
Resolved in #1199
